### ISSUE TITLE
Name Constraints: Discussion

### ISSUE THREAD

**Issue by lachellel** [uspki/policies#12]
Name Constraints

- Issuing certificates for US government properties
- Name constraints will include .gov and .mil 
- **any others?**

**Comment by LarryFrank** [comment]
DoD would want to include some specific .EDU domains.  The military academies and the national defense universities use .edu domain name for their web sites.  USCG (sorta DoD and DHS) and US Merchant Marine (not DoD) academies also use .edu.  There maybe other academic organizations within the Fed that do likewise.  Would need to canvas agencies to see.

**Comment by lachellel** [comment]
Referencing sections in BRD for discussion: 

- 7.1.5 Name Constraints: https://github.com/uspki/policies/blob/master/certificate-policy.md#715-name-constraints

**Comment by ab** [comment]
There might be a handful under `.fed.us`, not sure whether it makes sense to have those be in scope. These were pulled from pulse.cio.gov:

- http://fs.fed.us/
- http://www.csosa.fed.us/

**Comment by grandamp** [comment]
How about:  https://navy.com/ ?

**Comment by konklone** [comment]
One thing we should weigh when considering additional allowed names is not just whether it's a government property, but whether the cost (financial or security) of that government property continuing to use commercial certificates is larger than the complexity cost of supporting more names.

For example, right now name constraint support is not widely present among non-browser clients, and I _believe_ that in some browser clients, it's implemented in browser-land code rather than in a low-level library shared across the community (e.g. libssl or nss). 

So, complexity of implementation may well be a concern to the community when we discuss appropriate name constraints.

Since fed.us is on the way out, my immediate reaction is that we're better off not including it, and using this as an excuse to hasten its transition to the .gov space.

**Comment by buckaroogeek** [comment]
GSA, via the domain registrar, has already chosen to not implement DNSSEC
on the fed.us domain which is one of the incentives driving the FS to move
to fs.usda.gov. I would say, however, that even if we were to make the
change to fs.usda.gov tomorrow, there would still be a lengthy transition
period of years (5? 10?) where fs.fed.us names would need to remain viable
so that redirects continue to function. I therefore expect that there will
be a persistent requirement for TLS support for the fed.us name space
during that period.

**Comment by weirdscience** [comment]
DOD also has some .edu (service academies).

**Comment by konklone** [comment]
@weirdscience I've heard that a few times - at some point, someone will need to identify and list them. And the same thing will apply as I mentioned above -- weighing the complexity cost to the benefits of moving them away from commercial certificates.

**Comment by LarryFrank** [comment]
> DOD also has some .edu (service academies).

Yes...  True - also the National Defense Universities use .EDU

> weighing the complexity cost to the benefits of moving them away from commercial certificates

Talking to the Root team in DoD, they don't see a it as that big a problem, at least for the ones DoD needs.  The question I have is what we need to say in the policy on this topic at this point.

Larry

**Comment by weirdscience** [comment]
Two approaches:

1) Unconstrained TLD- require NPE issuers to have a webtrust for SSL audit with the FPKI addendum.

2) Constrained TLD - Only issue to .gov and .mil using government audit standard. Government websites outside this name space must use commercial provider. Policy would have to state the constraint.

Anything different than this would require individual root program contact and negotiation/exception.

**Comment by TLSrUS** [comment]
Is it possible to have weirdscience's 2) Constrained TLD above and include some specific .EDU domains used by military academies and the national defense universities as LarryFrank suggested above?

**Comment by konklone** [comment]
It's definitely possible! I think we just need to be prepared for community discussion about the complexity impact.

**Comment by pzb** [comment]
(community voice) For the described cases, it probably makes more sense to have an unconstrained roots and require every CA certificate issued by the root to have name constraints.  Then you can have a "Defense Education" issuer that has all the .edu name and other issuers for the simpler mil/gov case.

**Comment by techliaison** [comment]
It could also be handled with explicit permits in the name constraints if DoD didn't want separate issuing CAs.
Of course it is my understanding that not all browsers actually process name constraints but that is a different issue,

**Comment by LarryFrank** [comment]
>  It could also be handled with explicit permits in the name constraints if DoD didn't want separate issuing CAs.

I am told (and expect) that DoD would operate its own issuing CAs.  

> (community voice) For the described cases, it probably makes more sense to have an unconstrained roots and require every CA certificate issued by the root to have name constraints. Then you can have a "Defense Education" issuer that has all the .edu name and other issuers for the simpler mil/gov case.

And, my root engineer tells me that typically the Root is unconstrained and the issuing CAs have the constraints placed on them  But that isn't a policy question, in any event - that is implementation.  Note, the defense issuing CA would also have ".mil/.gov."

**Comment by lachellel** [comment]
My vote - let's make a consensus and put this to rest?  There are many more reviews!

- I vote for constraining in the **CP** so as not to overextend the reach 
- Realize .edu is in use
- Majority of .edu is not USG
- Every use case, we can find one-offs (.com, .edu)

**Comment by lachellel** [comment]
**Summarizing Name Constraints discussion points and approach:**

@techliaison 
> 2) Constrained TLD - Only issue to .gov and .mil using government audit standard. Government websites outside this name space must use commercial provider. Policy would have to state the constraint.

Reference #78 and #79 

- subordinate / issuing CAs have the constraints
- constraints will be specified in policies up front - to be applied to all via CP
  - consistency is a key element to maintain for USG
- .mil / .gov only
  - all others - including .us or .edu - have alternate options available to use

**Comment by lachellel** [comment]
Two additional questions from teams who are not all on GitHub : 

1. We disallow IPAddresses in general per various USG policies.  Are there any samples in public where IPAddresses are used? lint, crt, etc? 

2. How do various clients currently handle complex NameConstraints?

**Comment by lachellel** [comment]
**Info only** - DHS sites outside the .gov and .mil domains: 

- https://www.dhs.gov/dhs-sites-outside-gov-or-mil-domains

Listed per OMB policy.

**Comment by gerv** [comment]
If you are still considering putting name constraints in the root, consult the CA HARICA (members of the CA/Browser Forum) who tried this and I believe ran into some difficulties.

**Comment by dzacharo** [comment]
As Gerv said, we used name constraints in our 2011 Root CA Certificate. To paraphrase a common saying in the UK, "mind the dot"(!) in the dNSName permittedSubtrees. Our goal back in 2011 was to limit issuance to *.gr, *.net, *.eu, *.org. At the time, it did not make sense to leave the "gr", "net", "eu", "org" in the permittedSubtrees because noone should get a certificate with dNSName:gr and so on. So, we decided to have the following name constraints in the permittedSubtrees:
dNSName: .gr
dNSName: .net
and so on. 
Browsers didn't mind the leading dot but later on, we discovered that some implementations (openssl, gnutls) did not accept the leading dot in the dNSName. During our tests, openssl, gnutls did not handle name constraints at all and we mistakenly thought that they handle them as we expected. I don't want to expand too much on that because it was a long discussion about the semantics of RFC5280, domain labels, etc. 
To cut a long story short, if you plan on using Name Constraints in the Root Certificate, make sure **not to** prepend a dot in the dNSName.
